Spark zipnumcluster job

jt55401 opened the pull request:
This is a cc-pyspark version of the zipnum clustering job (without the use of mrjob framework)

- This may be problematic, but, I am not sorting the results quite as strictly as the existing job - my read of the zipnum document makes it sound like as long as your index file points to right place, we should be fine. Not sure if this will really work until I test it though.
- My test plan was going to be to run the index server locally and confirm I can get urls in the index, and also not in the index (ie: in between chunks)

jt55401 committed:
Add an alternate spark version of indexwarcsjob (without mrjob)

jt55401 committed:
fix so we properly can skip fully failed files and log them.

jt55401 committed:
feat: draft version of zipnum cluster job as spark (untested)

jt55401 committed:
docs: adding note about zipnum testing and behaviour

jt55401 commented:
also note: @sebastian-nagel  - I do have a "more perfect sorting" version which uses reservoir sampling in my local history - if we end up needing it, it's already done.

sebastian-nagel commented on zipnumcluster-ccpyspark.py:
```diff
@@ -0,0 +1,161 @@
+                    if len(current_chunk) >= chunk_size:
+                        # Compress and write chunk
+                        chunk_data = ''.join(current_chunk).encode('utf-8')
+                        compressed = z.compress(chunk_data)
```
Whenever a new chunk is started the zlib compressobj needs to be reset. Also need to flush. Should be:
```python
z = zlib.compressobj(6, zlib.DEFLATED, zlib.MAX_WBITS + 16)
compressed = z.compress(chunk_data) + z.flush()
```

sebastian-nagel commented on zipnumcluster-ccpyspark.py:
```diff
@@ -0,0 +1,161 @@
+                # Handle final chunk
+                if current_chunk:
+                    chunk_data = ''.join(current_chunk).encode('utf-8')
+                    compressed = z.compress(chunk_data) + z.flush()
```
Here as well: start a new compressobj.

sebastian-nagel commented on zipnumcluster-ccpyspark.py:
```diff
@@ -0,0 +1,161 @@
+                        
+                        # Create single index entry per record
+                        for sk, ts in chunk_records:
+                            index_entries.append((sk, ts, partition_id, current_offset, chunk_length))
```
The secondary index index (`cluster.idx`) only contains the offset for the first record of every zipnum block but not all of them.

jt55401 replied:
ah - this (and your explanation above) makes more sense to me now, thank you.
(I am still not seeing any of that in the docs I was able to find, but, I trust you, and will adjust)

sebastian-nagel commented on zipnumcluster-ccpyspark.py:
```diff
@@ -0,0 +1,161 @@
+                    f.write(compressed)
+                    
+                    for sk, ts in chunk_records:
+                        index_entries.append((sk, ts, partition_id, current_offset, chunk_length))
```
Also: only first record.

sebastian-nagel commented:
> not sorting the results quite as strictly as the existing job - my read of the zipnum document makes it sound like as long as your index file points to right place, we should be fine.

After thinking longer about it: some types of queries will still work but others don't. The problem starts when the software reading the CDX index assumes that it is totally sorted. This especially applies to any kind of range queries.

For example, the query
```
https://index.commoncrawl.org/CC-MAIN-2024-42-index?url=*.com&showNumPages=true&output=json
```
returns `{"pages": 84874, "pageSize": 5, "blocks": 424369}`.

So, for the `.com` top-level domain there are maximum `424369 * 3000 = 1273107000` or 1.27 billion captures in CC-MAIN-2024-42.

Basically this is counting the number of lines in the `cluster.idx` which start with `com,`:
```
$> grep -c '^com,' collections/CC-MAIN-2024-42/indexes/cluster.idx 
424368
```
Because there might be also results in the zipnum block before the first one, 1 is added to the number of lines.

If the zipnum blocks are non-contiguous, we'd need to add 1 for every contiguous range of block. Naturally, the result will become less precise.

In addition, there's more work to do for larger range queries. That's what the statement "Generally, this overhead [of the zipnum index] is negligible when looking up large indexes, and non-existent when doing a range query across many CDX lines." (<https://pywb.readthedocs.io/en/latest/manual/indexing.html#zipnum>)

On the other end, queries for single URLs might work the same and with the same performance independent from the partitioning scheme.

> a "more perfect sorting" version

What does it mean? [Total order sorting](https://blog.ditullio.fr/2016/01/04/hadoop-basics-total-order-sorting-mapreduce/)?

> My test plan was going to be to run the index server locally and confirm I can get urls in the index, and also not in the index (ie: in between chunks)

All kind of range queries also need to be tested:
- using a wildcard: `*.com`, `*.org.uk`, `youtu.be/a*`
- `matchType={prefix,host,domain}`
- [showNumPages](https://pywb.readthedocs.io/en/latest/manual/cdxserver_api.html#shownumpages-true)

Of course, even then: we'd need to document for our users the new CDX index sorting and spread this information. We do not know which assumptions are made in any third-party software and whether they rely on the total order sorting. This alone might make it less work to implement the total order sorting.

jt55401 commented:
> What does it mean? [Total order sorting](https://blog.ditullio.fr/2016/01/04/hadoop-basics-total-order-sorting-mapreduce/)?

I mentioned in next sentence fragment - I used same technique as was used in hadoop version - reservoir sampling to produce the ranges, then another pass using those ranges to do the shards.

I will find that version in my local history and check it when I work on this next.

sebastian-nagel commented:
> reservoir sampling to produce the ranges, then another pass using those ranges to do the shards.

Maybe it's not necessary to do the sampling step - Spark has a [sortBy](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.RDD.sortBy.html) (or sortByKey) method which does a total order sorting with N partitions. We use it to sort the vertices of the host-level webgraph before enumerating them. Same with the reservoir sampling: the partitions are not perfectly balanced, but the balance is acceptable.

Note: Spark has also methods to only sort the data within the partitions, they are usually named by "WithinPartitions", see for example [repartitionAndSortWithinPartitions](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.RDD.repartitionAndSortWithinPartitions.html).

jt55401 commented:
> > reservoir sampling to produce the ranges, then another pass using those ranges to do the shards.
> 
> Maybe it's not necessary to do the sampling step - Spark has a [sortBy](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.RDD.sortBy.html) (or sortByKey) method which does a total order sorting with N partitions. We use it to sort the vertices of the host-level webgraph before enumerating them. Same with the reservoir sampling: the partitions are not perfectly balanced, but the balance is acceptable.
> 
> Note: Spark has also methods to only sort the data within the partitions, they are usually named by "WithinPartitions", see for example [repartitionAndSortWithinPartitions](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.RDD.repartitionAndSortWithinPartitions.html).


Indeed - I was aware of these, but not all of them, and only some of the nuance. I've done some deep reading, and, by my best assessment:

my informal definition of "perfect sort" is the last record of 1 partition will be less than the first record of the next partition (so, if I go through the partitions in order, I will never get records out of order.)

* sort() (and sortByKey)- will work - and will sort perfectly. As you say, it does require a shuffle, so, it's "expensive" by comparison.
* repartitionAndSortWithinPartitions seems like the best option.
  * using hash to partition, like I do now - will be faster, as it will require fewer passes through the data, and will produce more uneven partitions, but, they should be "perfectly" sorted.
  * using reservoir sampling - requires a sampled pass through the data to generate ranges, and will generate more even partitions, with "almost" perfect sort - but not 100 due to sampling I think...
  * using dynamic reservoir sampling - will be about as fast as hash, but, the initial distributions to each partition will be relatively "messy" (I do have a version of this as well)

I'm leaning towards repartitionAndSortWithinPartitions, using hash of the url - but I may change my mind after running a few jobs and seeing how uneven they are... IMHO, 5-10% variance seems fine, if it's much more than that, it doesn't feel as good (though, that's why I want to read the zipnum code as I state below - it may not have a practical issue... so, could be fine)

Since I'm waiting on/monitoring other jobs anyway, I'm going to take another block of time tomorrow to do similar deep read of zipnum code, just so I have much better understanding of that as well. (specifically, I'm going to read the index server's code which USES zipnum, as that's the part that is still murky to me)

Thanks again for the input @sebastian-nagel , much appreciated.

wumpus commented:
Everyone's expectation is that the cdx index shards and parquet shards surt values should not overlap.

For cdx, that's expected by pywb.

For parquet, it's important for optimization. We do have a few parquet indexes for which that isn't true, and it's a problem we will fix someday.

jt55401 commented:
> Everyone's expectation is that the cdx index shards and parquet shards surt values should not overlap.

Got it - I think with hash and reservoir sampled approaches I outlined, they should not overlap at the shard level (as the former would not allow it, and the latter would be matching what we already do today pretty exactly). There MAY be overlapping gzip chunks that overlap (very small amounts with reservoir, and potentially rather larger amounts with hash) - but, as long as the secondary index reflects those properly, I don't think it'll be an issue based on my read of the index server side of things.

jt55401 commented:
I will get back to this task Monday, so, plenty of time to discuss if I'm wrong there... I will bring it up on eng. call, and if we need to talk, we can do it then.

sebastian-nagel commented:
>> Everyone's expectation is that the cdx index shards and parquet shards surt values should not overlap.

> Got it - I think with hash and reservoir sampled approaches I outlined, they should not overlap at the shard level

What Greg means is that there must be zero overlap for all ranges defined by the first and the last SURT in a zipnum block. It's important because the secondary index (cluster.idx) only stores the first SURT but not the last one. But with strict sorting, the last one must be lower (sorts before) than the first one of the next zipnum block.

For Parquet zero overlap is an optimization but no requirement: every Parquet file and row group has the min **and max** values in the statistics in the footer.

jt55401 committed:
feat: reservoir sampled method of zipnum cluster job

jt55401 committed:
fix: simpler ordered partition id

jt55401 committed:
fix: re-init and flush gz

jt55401 committed:
fix: bugfix for edge conditions

jt55401 committed:
fix: getting order and structure of final index proper, and in single file

jt55401 committed:
fix: fix cdx filenames

jt55401 committed:
fix: final cluster.idx logic

jt55401 commented:
I have updates to this task in another branch - for now, I'm going to preserve the existing approach, reservoir, and get this task finished. I have it all working locally and will be testing tonight/tomorrow in s3 on a full crawl, and will then re-do the PR to reflect (I'll probably merge it into this branch, to keep things simple, and preserve the above history.

jt55401 committed:
fix: get zipnumcluster-pyspark working with s3/emr

jt55401 committed:
chore: merged latest normal spark solution upwards

jt55401 committed:
chore: merging other changes

jt55401 committed:
fix: numerous bugfixes from complex merge issues

jt55401 committed:
fix: more emr bugfixes and tweaks

jt55401 committed:
fix: temporary working version  of zipnum job using normal spark features (need to refactor ccpyspark back into it now)

jt55401 committed:
chore: going back to using CCFileProcessorSparkJob

jt55401 committed:
fix: bugfixes found from integration testing

sebastian-nagel commented on zipnumcluster-ccpyspark.py:
```diff
@@ -0,0 +1,263 @@
+        parser.add_argument("--output_base_url", required=False,
+                            default='my_cdx_bucket',
+                            help="destination for output")
+        parser.add_argument("--partition_boundries_file", required=False,
```
If `--partition_boundries_file` is not specified, the job fails:
```
Traceback (most recent call last):
  File ".../webarchive-indexing/zipnumcluster-ccpyspark.py", line 263, in <module>
    job.run()
  File ".../webarchive-indexing/sparkcc.py", line 187, in run
    self.run_job(session)
  File ".../webarchive-indexing/zipnumcluster-ccpyspark.py", line 238, in run_job
    self.write_output_file(boundries_file_uri, f)
  File ".../webarchive-indexing/sparkcc.py", line 840, in write_output_file
    uri_match = self.data_url_pattern.match(uri)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: expected string or bytes-like object, got 'NoneType'
```

Is this a required argument?

jt55401 replied:
Yes - it's either where the file is written to, or read from (if it already exists)

sebastian-nagel commented on zipnumcluster_cc_pyspark.py:
```diff
@@ -0,0 +1,263 @@
+        # loop over the output files and concatenate them into a single final file
+        with open('cluster.idx', 'wb') as f:
+            for idx_file,_ in rdd:
+                with self.fetch_file(output_base_url + idx_file) as idx_fd:
```
The default of `--output_base_url` is `my_cdx_bucket` (no trailing slash). So, when testing locally, I get the error:
```
4/12/20 14:03:37 INFO ZipNumClusterCdx: Reading local file my_cdx_bucketidx-00000.idx
Traceback (most recent call last):
  File ".../webarchive-indexing/zipnumcluster-ccpyspark.py", line 263, in <module>
    job.run()
  File ".../webarchive-indexing/sparkcc.py", line 187, in run
    self.run_job(session)
  File ".../webarchive-indexing/zipnumcluster-ccpyspark.py", line 252, in run_job
    with self.fetch_file(output_base_url + idx_file) as idx_fd:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../webarchive-indexing/sparkcc.py", line 727, in fetch_file
    warctemp = open(uri, 'rb')
               ^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '.../webarchive-indexing/my_cdx_bucketidx-00000.idx'
```

Looks like, for testing locally, `--output_base_url` needs to be of the form: `file:/absolute/path/`.

jt55401 replied:
yes, that's correct (and is shown in the integration tests/scripts in the cc-redact project)

sebastian-nagel commented on zipnumcluster_cc_pyspark.py:
```diff
@@ -0,0 +1,263 @@
+                            help="number of partitions/shards")
+
+    def run_job(self, session):
+        os.makedirs(self.args.output_base_url, exist_ok=True)
```
See below: if `--output_base_url` is `file:/absolute/path/`, on Linux a relative folder `file:` is created in the current directory.

jt55401 replied:
I will double check that - it's not happening for me, but I may have omitted file:, so, may be testing it differently

sebastian-nagel reviewed (changes requested):
There are some hick-ups if trying to create a zipnum index locally.

Finally, I was able to create a zipnum index by setting
-  `--output_base_url=file:/absolute/path/to/index/`
  - but this creates a directory tree `file:/absolute/path/to/index/` in the current directory
- `--partition_boundries_file="/absolute/path/to/index/cluster.idx`
  - please, also fix the typo "boundries" (should be "boundaries")

The created index, both `cdx-*.gz` and `cluster.idx`, look good:
- all records from the input are in the zipnum index
- no sorting issues found
- offsets in cluster.idx point to valid zipnum blocks

sebastian-nagel commented on zipnumcluster-ccpyspark.py:
```diff
@@ -0,0 +1,263 @@
+        rdd = rdd.repartitionAndSortWithinPartitions(
+            numPartitions=num_partitions,
+            partitionFunc=lambda k: get_partition_id(k,boundaries),
+            keyfunc=lambda x: x[0]) \
```
This will sort on the first character of the SURT key only. Should use the identity function (just leave the param `keyfunc` away).

jt55401 replied:
humn - OK, I had put that in there because I thought x was a tuple at that point, and [0] was the key - but based on what you're saying, does keyfunc already get passed the first element of the tuple? I will double check the docs, but, it sounds reasonable.

jt55401 replied:
yup, right you are:

this is from spark source:
```
key=lambda k_v: keyfunc(k_v[0])
```

sebastian-nagel replied:
That's a subtle difference between [sortByKey](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.RDD.sortByKey.html) and [sortBy](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.RDD.sortBy.html). And of course, from the documentation it's even less clear how [repartitionAndSortWithinPartitions()](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.RDD.repartitionAndSortWithinPartitions.html) does behave. Ok, in Scala or Java this would just result in an Exception "Expected list or tuple, got string". With Python this is tricky.

jt55401 committed:
fix: spelling of boundaries

jt55401 committed:
fix: marking args as required

jt55401 committed:
fix: remove keyfunc, it isn't needed (and was causing issue that sort was only working on first char

jt55401 committed:
chore: adding TODO for upcoming tasks

jenenglish commented:
@sebastian-nagel This is another open PR related to mrjob.  Should we keep open, close, reassign?

sebastian-nagel committed:
fix: code formatting and comments

sebastian-nagel commented:
This PR including #7 still require:
- testing including index validation
- removal of code duplicated from CCFileProcessorSparkJob
- few TODOs in code to be addressed
- squash commits before merging

sebastian-nagel commented:
A first small test run with verification surfaced one issue:

The cdx-NNNNN.gz files are sorted by SURT, but not by capture time as secondary sorting criterion.
- This is different from the MRJob which sorts lines, that is <SURT, timestamp, JSON>
- If there are many captures for the same SURT, a wayback implementation may fail to pick the right timestamp if these are not sorted.

sebastian-nagel committed:
fix: sort by SURT key *and* capture timestamp

sebastian-nagel committed:
refactor: rename Python files to valid module names

sebastian-nagel committed:
refactor: move functions out of global scope

- make all functions in the global scope static methods
  of the class ZipNumClusterCdx. Static methods are
  required by Spark serialization if methods are used
  in RDD methods (map, partition, etc.)

fix: make methods requiring serialization static

sebastian-nagel committed:
fix: do not create output directory tree in local filesystem

sebastian-nagel committed:
fix: type error when using partition boundary file
     passed per --partition_boundaries_file

sebastian-nagel committed:
refactor: code documentation, formatting and cleanup

sebastian-nagel committed:
fix: use correct timestamp in cluster.idx

sebastian-nagel committed:
refactor: persistence and temporary files

- remove commentented code to persist RDD

- write per-shard indexes to temporary location,
  only the cluster.idx is written to the final location

sebastian-nagel committed:
fix: unify logging, address Pylint warnings

sebastian-nagel committed:
fix: suppress ignored CLI options

sebastian-nagel committed:
Spark jobs to build index: README, add shell script and requirements

sebastian-nagel commented:
Reviewed and reworked code. Successfully tested both locally and on a Hadoop cluster (Spark on YARN) on 1% of the index records of a monthly crawl.
- script to run the jobs on a cluster: `run_index_ccpyspark.sh`
- script to verify a ZipNum index: [verify_zipnum.sh](https://github.com/user-attachments/files/23634683/verify_zipnum.sh)

sebastian-nagel reviewed (approved):
All points resolved. Successfully tested, both locally and on a cluster.

sebastian-nagel committed:
refactor: adapt publish_index.sh for Spark jobs

sebastian-nagel committed:
refactor: increase driver memory and max. result size

Increase Spark driver memory to 6 GiB and
max. result size to 4 GiB in order to
address potential out-of-memory issues
when sampled URL index keys (SURT and timestamp)
are sent to the driver.

sebastian-nagel commented:
Spark zipnumcluster job is now running in production.
